python/ray/_raylet.pyx

```diff
@@ -481,6 +482,58 @@ cdef raise_if_dependency_failed(arg):
         raise arg
 
 
+cdef c_bool determine_if_retryable(
+    Exception e,
+    const c_string serialized_retry_exception_allowlist,
+    FunctionDescriptor function_descriptor,
+):
+    """Determine if the provided exception is retryable, according to the
+    (possibly null) serialized predicate function.
+
+    If the serialized predicate function is an empty string or is None once
+    deserialized, the exception is considered retryable and we return True.
+
+    This method can raise exceptions in several cases:
+        - Deserialization fails (TypeError)
+        - Deserialized predicate function is not a function (TypeError)
+        - Predicate function raises an error (RuntimeError)
+        - Predicate function output is not a boolean (TypeError)
+    """
+    if len(serialized_retry_exception_allowlist) == 0:
+        # No exception allowlist specified, default to all retryable.
+        return True
+
+    # Deserialize exception allowlist and check that e is in the allowlist.
+    try:
+        exception_allowlist = ray_pickle.loads(
+            serialized_retry_exception_allowlist,
+        )
+    except TypeError as inner_e:
+        # Exception allowlist deserialization failed.
+        msg = (
+            "Could not deserialize the retry exception allowlist "
+            f"function for task {function_descriptor.repr}. "
+            "Check "
+            "https://docs.ray.io/en/master/ray-core/objects/serialization.html#troubleshooting " # noqa
+            "for more information.")
+        raise TypeError(msg) from inner_e
+
+    if exception_allowlist is None:
+        # No exception allowlist specified, default to all retryable.
+        return True
+
+    if not isinstance(exception_allowlist, tuple):
```

Comment:
I think this should be an assert. We already validate the user inputs so this should never happen.

Reply:
Yeah this should be an assertion, although this is more about confirming that we convert the `List[Exception]` that the user provides to a `Tuple[Exception, ...]`.